### PR TITLE
Viewer for binary data

### PR DESCRIPTION
**Pull request by hajdam**
This introduces hexadecimal viewer for binary data based on BinEd (formely deltahex) library.

This is suppose to be sequence of separate smaller changes which can be easily read and considered/discussed.

- Added OSGI service for file viewer (and stub for file editor)
- Added OSGI service for configuration snapshots
- Text and image viewers/editor moved to separate modules
- Added new module for binary viewer (no editor yet) - viewer is used when viewed file is neither detected as text or image

Next steps should include refactoring viewers to use interface instead of abstract class + support switching between different viewers in single window.

### CODE REVIEW

File: mucommander-viewer-text/src/main/java/com/mucommander/viewer/text/TextViewer.java [hunk 1]
```diff
@@ -106,6 +105,7 @@ public void setFrame(final FileFrame frame) {
 
         getInputMap(JComponent.WHEN_IN_FOCUSED_WINDOW).put(KeyStroke.getKeyStroke(KeyEvent.VK_M, ActionEvent.CTRL_MASK), CUSTOM_FULL_SCREEN_EVENT);
     	getActionMap().put(CUSTOM_FULL_SCREEN_EVENT, new AbstractAction() {
+                @Override
```
Comment (ahadas) [inline mucommander-viewer-text/src/main/java/com/mucommander/viewer/text/TextViewer.java]: indentation?

File: mucommander-core/src/main/java/com/mucommander/osgi/FileViewerServiceTracker.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        LOGGER.info("FileViewerService is unregistered: " + service);
+    }
+
+    private static void addViewerService(FileViewerService service) {
+        int index = 0;
+        int orderPriority = service.getOrderPriority();
+        while (index < SERVICES.size() && orderPriority < SERVICES.get(index).getOrderPriority())
+        {
+            index++;
+        }
```
Comment (ahadas) [inline mucommander-core/src/main/java/com/mucommander/osgi/FileViewerServiceTracker.java]: how about sorting the list in `getViewerServices` instead?

File: mucommander-core/src/main/java/com/mucommander/snapshot/SnapshotHandler.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+package com.mucommander.snapshot;
+
+import com.mucommander.commons.conf.Configuration;
+
+/**
+ * Snapshot support for modules.
+ *
+ * @author Miroslav Hajda
+ */
+public interface SnapshotHandler {
```
[the remark this answers is not in the document]
Comment (hajdam) [reply mucommander-core/src/main/java/com/mucommander/snapshot/SnapshotHandler.java]: I guess I'll go with "MuSnapshotable" if it's ok.

File: mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryViewer.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        editMenu = new JMenu(Translator.get("text_viewer.edit"));
+        MnemonicHelper menuItemMnemonicHelper = new MnemonicHelper();
+
+        copyItem = MenuToolkit.addMenuItem(editMenu, Translator.get("text_viewer.copy"), menuItemMnemonicHelper, null, this);
+        copyItem.addActionListener(new ActionListener() {
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                binaryViewerImpl.copy();
+            }
+        });
```
[the remark this answers is not in the document]
Comment (hajdam) [reply mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryViewer.java]: I was not sure if Java 8 / lambda is allowed. Did I maybe miss it mentioned in wiki somewhere?

[the remark this answers is not in the document]
Comment (ahadas) [reply mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryViewer.java]: well, maybe not all the relevant wiki pages are up-to-date but the [README page](https://github.com/mucommander/mucommander/blob/master/README.md) states that Java 8 is required for development (this should actually be updated now to [Java 11](https://github.com/mucommander/mucommander/blob/master/build.gradle#L27)..). feel free to add/suggest adding it elsewhere

File: mucommander-core/src/main/java/com/mucommander/osgi/FileViewerServiceTracker.java (hunk 2 above)
Comment (ahadas) [reply mucommander-core/src/main/java/com/mucommander/osgi/FileViewerServiceTracker.java]: on second thought, I see `getViewerServices` is called each time a file is opened for viewing so this way is better. but maybe, for simplicity, we can just call Collections#sort with an appropriate Comparator?

File: mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryViewer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (hajdam) [reply mucommander-viewer-binary/src/main/java/com/mucommander/viewer/binary/BinaryViewer.java]: I used brackets / braces, but I can convert it to single line, if it's preferable.

File: mucommander-viewer-text/src/main/java/com/mucommander/viewer/text/TextViewer.java (hunk 1 above)
Comment (hajdam) [reply mucommander-viewer-text/src/main/java/com/mucommander/viewer/text/TextViewer.java]: I autoformated class by mistake, but I hope it's ok. Btw. I'm resolving conversations, not sure if that is how to mark requested change was done. (This is actually my second PR on GitHub)

Comment (ahadas) [reply mucommander-viewer-text/src/main/java/com/mucommander/viewer/text/TextViewer.java]: formatting looks good now, thanks. as for resolving conversations, what I usually do is replying Done like on Gerrit and let the commenter resolve the conversation - I'm not sure if there's a "right way" or a standard way for this though.

**Comment by ahadas** [comment]
@hajdam  thanks for this contribution!

**Pull request merged**